Armadillo/src/main/java/com/scribd/armadillo/download/drm/ArmadilloDrmSessionManagerProvider.kt

```diff
@@ -0,0 +1,136 @@
+    // These fields identical to DefaultDrmSessionManagerProvider //
+
+    private val lock = Any()
+
+    @GuardedBy("lock")
+    private var drmConfiguration: DrmConfiguration? = null
+
+    @GuardedBy("lock")
+    private var manager: DrmSessionManager? = null
+    private val drmHttpDataSourceFactory: DataSource.Factory? = null
+    private val userAgent: String? = null
+
+    // End DefaultDrmSessionManagerProvider fields //
+    /////////////////////////////////////////////////
+
+    companion object {
+        const val TAG = "ArmadilloDrmSMProvider"
+    }
+
+    /**  Identical to DefaultDrmSessionManagerProvider method, */
+    override fun get(mediaItem: MediaItem): DrmSessionManager {
+        Assertions.checkNotNull(mediaItem.localConfiguration)
+        val drmConfiguration = mediaItem.localConfiguration!!.drmConfiguration
+        if (drmConfiguration != null && Util.SDK_INT >= 18) {
+            synchronized(this.lock) {
+                if (!Util.areEqual(drmConfiguration, this.drmConfiguration)) {
+                    this.drmConfiguration = drmConfiguration
+                    this.manager = this.createManager(drmConfiguration)
+                }
+                return Assertions.checkNotNull<DrmSessionManager?>(this.manager)
+            }
+        } else {
+            return DrmSessionManager.DRM_UNSUPPORTED
+        }
+    }
+
+    /** Near identical to DefaultDrmSessionManagerProvider method, except for the indicated line */
+    private fun createManager(drmConfiguration: DrmConfiguration): DrmSessionManager {
+        val dataSourceFactory = this.drmHttpDataSourceFactory
+            ?: DefaultHttpDataSource.Factory().setUserAgent(this.userAgent)
+        val httpDrmCallback = HttpMediaDrmCallback(if (drmConfiguration.licenseUri == null) null else drmConfiguration.licenseUri.toString(), drmConfiguration.forceDefaultLicenseUri, (dataSourceFactory)!!)
+        val var4: UnmodifiableIterator<*> = drmConfiguration.licenseRequestHeaders.entries.iterator()
+
+        while (var4.hasNext()) {
+            val entry: Map.Entry<String, String> = var4.next() as Map.Entry<String, String>
+            httpDrmCallback.setKeyRequestProperty(entry.key, entry.value)
+        }
+
+        val drmSessionManager = DefaultDrmSessionManager.Builder()
+            //this line is different, changing the DrmProvider
+            .setUuidAndExoMediaDrmProvider(drmConfiguration.scheme, ArmadilloDrmProvider(stateStore))
+            .setMultiSession(drmConfiguration.multiSession)
+            .setPlayClearSamplesWithoutKeys(drmConfiguration.playClearContentWithoutKey)
+            .setUseDrmSessionsForClearContent(*Ints.toArray(drmConfiguration.forcedSessionTrackTypes))
+            .build(httpDrmCallback)
+        drmSessionManager.setMode(DefaultDrmSessionManager.MODE_PLAYBACK, drmConfiguration.keySetId)
+        return drmSessionManager
+    }
+
+    /** New original provider for this class */
```

Comment:
The only thing this does differently is emit events to the dispatcher, right? Can you make that explicit in the docstring